gnovm/pkg/gnolang/debugger.go

```diff
@@ -0,0 +1,758 @@
+	ds, ok := st.(*defaultStore)
+	return ok && ds.iavlStore.Has([]byte(backendPackagePathKey(pkgPath)))
+}
+
+func fileContent(st Store, pkgPath, name string) (string, error) {
+	if isMemPackage(st, pkgPath) {
+		return st.GetMemFile(pkgPath, name).Body, nil
+	}
+	buf, err := os.ReadFile(name)
+	return string(buf), err
+}
+
+func linesAround(src string, index, n int) ([]string, int) {
+	lines := strings.Split(src, "\n")
+	start := max(1, index-n/2) - 1
+	end := min(start+n, len(lines))
+	if start >= end {
+		start = max(1, end-n)
+	}
+	return lines[start:end], start + 1
+}
+
+// ---------------------------------------
+const (
+	printUsage = `print|p <expression>`
+	printShort = `Print a variable or expression.`
+)
+
+func debugPrint(m *Machine, arg string) (err error) {
+	if arg == "" {
+		return errors.New("missing argument")
+	}
+	// Use the Go parser to get the AST representation of print argument as a Go expresssion.
+	ast, err := parser.ParseExpr(arg)
+	if err != nil {
+		return err
+	}
+	tv, err := debugEvalExpr(m, ast)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintln(m.Debugger.out, tv)
+	return nil
+}
+
+// debugEvalExpr evaluates a Go expression in the context of the VM and returns
+// the corresponding typed value, or an error.
+// The supported expression syntax is a small subset of Go expressions:
+// basic literals, identifiers, selectors, index expressions, or a combination
+// of those are supported, but none of function calls, arithmetic, logic or
+// assign operations, type assertions of convertions.
+// This is sufficient for a debugger to perform 'print (*f).S[x][y]' for example.
+func debugEvalExpr(m *Machine, node ast.Node) (tv TypedValue, err error) {
+	switch n := node.(type) {
+	case *ast.BasicLit:
+		switch n.Kind {
+		case token.INT:
+			i, err := strconv.ParseInt(n.Value, 0, 0)
+			if err != nil {
+				return tv, err
+			}
+			return typedInt(int(i)), nil
+		case token.CHAR:
+			r, _, _, err := strconv.UnquoteChar(n.Value[1:len(n.Value)-1], 0)
+			if err != nil {
+				return tv, err
+			}
+			return typedRune(r), nil
+		case token.STRING:
+			s, err := strconv.Unquote(n.Value)
+			if err != nil {
+				return tv, err
+			}
+			return typedString(s), nil
+		}
+		return tv, fmt.Errorf("invalid basic literal value: %s", n.Value)
+	case *ast.Ident:
+		if tv, ok := debugLookup(m, n.Name); ok {
+			return tv, nil
+		}
+		return tv, fmt.Errorf("could not find symbol value for %s", n.Name)
+	case *ast.ParenExpr:
+		return debugEvalExpr(m, n.X)
+	case *ast.StarExpr:
+		x, err := debugEvalExpr(m, n.X)
+		if err != nil {
+			return tv, err
+		}
+		pv, ok := x.V.(PointerValue)
+		if !ok {
+			return tv, fmt.Errorf("Not a pointer value: %v", x)
+		}
+		return pv.Deref(), nil
+	case *ast.SelectorExpr:
+		x, err := debugEvalExpr(m, n.X)
+		if err != nil {
+			return tv, err
+		}
+		// TODO: handle selector on package.
+		tr, _, _, _, _ := findEmbeddedFieldType(x.T.GetPkgPath(), x.T, Name(n.Sel.Name), nil)
```

Comment:
We used to have a linter for this called "[dogsled](https://github.com/alexkohler/dogsled)" 😄